### PR TITLE
Decimal precision/scale and string length

### PR DESCRIPTION
**Pull request by jseldess**
Update decimal docs to cover precision and scale.

Update string docs to cover length limit. 

In response to https://github.com/cockroachdb/cockroach/pull/5750

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/200)

### CODE REVIEW

**Comment by knz** [comment]
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[decimal.md, line 19 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/200#-KEf4n4PFq4Uy4-dJK0Z:-KEf4n4QFEdLmFxWzwy1:1792083149) ([raw file](https://github.com/cockroachdb/docs/blob/8ccb46c847d367323529ff23e847101a36fe2285/decimal.md#L19)):*
This is correct.
You could also mention that a column declared with `DECIMAL(precision)` is equivalent to `DECIMAL(precision, 0)`.

---

*[string.md, line 10 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/200#-KEf5YyMDfcVxs88v6T2:-KEf5YyMDfcVxs88v6T3:444566982) ([raw file](https://github.com/cockroachdb/docs/blob/8ccb46c847d367323529ff23e847101a36fe2285/string.md#L10)):*
This is slightly incorrect. `CHAR`, `VARCHAR` and `TEXT` are aliases for `STRING`. `CHAR(n)` and `VARCHAR(n)` are aliases for `STRING(n)`.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/200)*

File: decimal.md [hunk 1]
```diff
@@ -14,7 +14,9 @@ In CockroachDB, the following are aliases for `DECIMAL`:
 
 ## Precision and Scale
 
-Precision (the maximum count of digits in a whole number, both to the left and right of the decimal point) and scale (the maximum count of digits to the right of the decimal point) are always unlimited for a `DECIMAL` value. If you specify these limits by declaring `DECIMAL(precision, scale)` on a column, they will not be enforced. 
+To limit a decimal column, use `DECIMAL(precision, scale)`, where `precision` is the maximum count of digits both to the left and right of the decimal point, and `scale` is the maximum count of digits to the right of the decimal point.
+
+When inserting a decimal value, if the digits to the right of the decimal point exceed the column's `scale`, CockroachDB rounds to the scale. After rounding, if the total count of digits exceeds the column's `precision`, CockroachDB gives an error.
```
Comment (petermattis) [inline decimal.md]: The scale is the precise count of digits to the right of the decimal point. That is, `DECIMAL(5,4)` will always have exactly 4 digits to the right of the decimal point.

```
pmattis=# select 1::decimal(5,4);
 numeric
---------
  1.0000
```

We don't seem to handle this correctly yet.

File: string.md [hunk 2]
```diff
@@ -17,9 +17,9 @@ In CockroachDB, the following are aliases for `STRING`:
 
 ## Length
 
-At this time, length is always variable for a `STRING` value. If you specify a fixed length by declaring `CHAR(n)` or `VARCHAR(n)` on a column, it will not be enforced.
+To limit the length of a string column, use `CHAR(n)` or `VARCHAR(n)`, where `n` is the maximum number of characters allowed.
 
-{{site.data.alerts.callout_info}}A future version of CockroachDB will support enforcing length limits on strings.{{site.data.alerts.end}}
+When inserting a string value, if the value exceeds the column's length limit, Cockroach gives an error.
```
Comment (petermattis) [inline string.md]: You should also note that when casting to a string with a length limit, truncation is performed:

```
root@:26257> select 'hello world'::char(5);
+--------------------------------+
| CAST('hello world' AS CHAR(5)) |
+--------------------------------+
| hello                          |
+--------------------------------+
```

File: string.md [hunk 3]
```diff
@@ -28,16 +28,16 @@ When declaring a `STRING`, format it as `'a1b2c3'`.
...
-+-------+--------+-------+---------+
-| a     | STRING | false | NULL    |
-| b     | STRING | true  | NULL    |
-| c     | STRING | true  | NULL    |
-+-------+--------+-------+---------+
++-------+-----------+-------+---------+
+| Field |  Type     | Null  | Default |
++-------+-----------+-------+---------+
+| a     | STRING    | false | NULL    |
+| b     | STRING(6) | true  | NULL    |
```
Comment (petermattis) [inline string.md]: Heh, we don't actually support `STRING(N)`, only `{CHARACTER,CHAR,VARCHAR}(n)`. This is a bug. Mind filing an issue.

**Comment by petermattis** [comment]
LGTM

**Comment by jseldess** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[decimal.md, line 19 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/200#-KEcdaHqsCUuSOnN9XNi-r1-19:-KEg-vjpdNwJLy75eW6x:1943849619) ([raw file](https://github.com/cockroachdb/docs/blob/8ccb46c847d367323529ff23e847101a36fe2285/decimal.md#L19)):*
I see. Correcting with next commit. Looks to me like we truncate to the specified scale as long as the total digits don't exceed the specified precision. But if you're "less than" scale, we don't fill in with 0s. Is that what you mean by not handling this correctly yet?

---

*[decimal.md, line 19 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/200#-KEf4n4PFq4Uy4-dJK0Z:-KEgJgAqgrZNj0ISzEdH:-1218437995) ([raw file](https://github.com/cockroachdb/docs/blob/8ccb46c847d367323529ff23e847101a36fe2285/decimal.md#L19)):*
Adding that note with next commit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/200)*

**Comment by jseldess** [comment]
*[string.md, line 10 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/200#-KEf5YyMDfcVxs88v6T2:-KEgR8eqiqgDJlJM50AB:-1545094195) ([raw file](https://github.com/cockroachdb/docs/blob/8ccb46c847d367323529ff23e847101a36fe2285/string.md#L10)):*
This is a good point, Raphael, but I'm a little confused about how to document this.  Currently, `STRING(n)` isn't even allowed:  

```
root@:26257> create table strings (a string, b string(n));
pq: syntax error at or near "("
create table strings (a string, b string(n));
```

Isn't it a little strange to say that `char(n)` and `varchar(n)` are aliases for  a form that isn't allowed? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/200)*

**Comment by jseldess** [comment]
*[string.md, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/200#-KEcdaHqsCUuSOnN9XNj-r1-38:-KEgRsTZRNZw2ho7D9cL:-1430866942) ([raw file](https://github.com/cockroachdb/docs/blob/8ccb46c847d367323529ff23e847101a36fe2285/string.md#L38)):*
Ah, yes, just added a comment related to this. I'll file a bug.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/200)*

**Comment by jseldess** [comment]
*[string.md, line 10 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/200#-KEf5YyMDfcVxs88v6T2:-KEgS0SvgYZKo8H9tIvR:-1545062531) ([raw file](https://github.com/cockroachdb/docs/blob/8ccb46c847d367323529ff23e847101a36fe2285/string.md#L10)):*
Just saw Peter's comment below that this is a bug. I'll file that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/200)*

**Comment by jseldess** [comment]
*[string.md, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/200#-KEcdaHqsCUuSOnN9XNj-r1-38:-KEgTkJ9GzRi27ZYVmP-:-1989392754) ([raw file](https://github.com/cockroachdb/docs/blob/8ccb46c847d367323529ff23e847101a36fe2285/string.md#L38)):*
Bug filed: https://github.com/cockroachdb/cockroach/issues/5892

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/200)*

**Comment by jseldess** [comment]
*[string.md, line 22 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/200#-KEcdaHqsCUuSOnN9XNj-r1-22:-KEg_r8OhTuDCZwmkHnI:1028098577) ([raw file](https://github.com/cockroachdb/docs/blob/8ccb46c847d367323529ff23e847101a36fe2285/string.md#L22)):*
Noted with next commit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/200)*

**Comment by jseldess** [comment]
*[string.md, line 10 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/200#-KEf5YyMDfcVxs88v6T2:-KEg_ww9xHiGwFO7HNX8:425882303) ([raw file](https://github.com/cockroachdb/docs/blob/8ccb46c847d367323529ff23e847101a36fe2285/string.md#L10)):*
I'll update the `Length` section to suggest using `string(n)` once the bug's resolved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/200)*

**Comment by knz** [comment]
:lgtm: 

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/200#-:-KEgdBROjEihtPB4XfY6:980077423)*

**Pull request merged**